federationtypes.go

```diff
@@ -4,10 +4,100 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"unicode"
 
 	"github.com/matrix-org/util"
 )
 
+// A ServerName is the name a matrix homeserver is identified by.
+// It is a DNS name or IP address optionally followed by a port.
+//
+// https://matrix.org/docs/spec/appendices.html#server-name
+type ServerName string
+
+// ParseAndValidateServerName splits a ServerName into a host and port part,
+// and checks that it is a valid server name according to the spec.
+//
+// if there is no explicit port, returns '-1' as the port.
+func ParseAndValidateServerName(serverName ServerName) (host string, port int, valid bool) {
+	host, port = splitServerName(serverName)
+	valid = false
+
+	// the host part must be one of:
+	//  - a valid (ascii) dns name
+	//  - an IPv4 address
+	//  - an IPv6 address
+
+	if host[0] == '[' {
+		// must be a valid IPv6 address
+		if host[len(host)-1] != ']' {
+			return
+		}
+		ip := host[1 : len(host)-1]
+		if net.ParseIP(ip) == nil {
+			return
+		}
+	} else if unicode.IsDigit(rune(host[0])) {
```

Comment:
gah; this is completely bogus. Hostnames can begin with digits.

Reply:
Oh poo, they can: https://0net.io/